src/main/java/hudson/plugins/promoted_builds/conditions/GroovyCondition.java

```diff
@@ -0,0 +1,121 @@
+package hudson.plugins.promoted_builds.conditions;
+
+import groovy.lang.Binding;
+import hudson.EnvVars;
+import hudson.Util;
+import hudson.model.AbstractBuild;
+import hudson.plugins.promoted_builds.PromotionBadge;
+import hudson.plugins.promoted_builds.PromotionCondition;
+import hudson.plugins.promoted_builds.PromotionProcess;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SecureGroovyScript;
+import org.jenkinsci.plugins.scriptsecurity.scripts.UnapprovedClasspathException;
+import org.jenkinsci.plugins.scriptsecurity.scripts.UnapprovedUsageException;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.CheckForNull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Allow specification of Groovy scripts to qualify builds
+ */
+public class GroovyCondition extends PromotionCondition {
+    private static final Logger LOGGER = Logger.getLogger(GroovyCondition.class.getName());
+
+    private final @CheckForNull String unmetQualificationLabel;
+    private final @CheckForNull String metQualificationLabel;
+    private final SecureGroovyScript script;
+
+    @DataBoundConstructor
+    public GroovyCondition(final SecureGroovyScript script, final String unmetQualificationLabel, final String metQualificationLabel) {
+        this.unmetQualificationLabel = Util.fixEmptyAndTrim(unmetQualificationLabel);
+        this.metQualificationLabel = Util.fixEmptyAndTrim(metQualificationLabel);
+        this.script = script.configuringWithNonKeyItem();
+    }
+
+    // exposed for config.jelly
+    public SecureGroovyScript getScript() {
+        return script;
+    }
+
+    // exposed for config.jelly
+    public String getUnmetQualificationLabel() {
+        return unmetQualificationLabel;
+    }
+
+    // exposed for config.jelly
+    public String getMetQualificationLabel() {
+        return metQualificationLabel;
+    }
+
+    // exposed for promotion status page
+    public String getDisplayLabel() {
+        return unmetQualificationLabel == null ? Messages.GroovyCondition_UnmetQualificationLabel() : unmetQualificationLabel;
+    }
+
+    @Override
+    public PromotionBadge isMet(final PromotionProcess promotionProcess, final AbstractBuild<?, ?> build) {
+        final ClassLoader classLoader = Jenkins.getInstance().getPluginManager().uberClassLoader;
+        final Binding binding = new Binding();
+        binding.setVariable("promotionProcess", promotionProcess);
+        binding.setVariable("build", build);
+        Object result = null;
+        try {
+            result = script.evaluate(classLoader, binding);
+        } catch (final RejectedAccessException e) {
+            LOGGER.log(Level.WARNING, "Sandbox exception", e);
```

Comment:
I'm not sure about returning anything here since the API documentation mentions it should be done only when the condition is met and there is no mention about exceptions handling:
```
    /**
     * Checks if the promotion criteria is met.
     *
     * @param promotionProcess
     *      The promotion process being evaluated for qualification
     * @param build
     *      The build for which the promotion is considered.
     * @return
     *      non-null if the promotion condition is met. This object is then recorded so that
     *      we know how a build was promoted.
     *      Null if otherwise, meaning it shouldn't be promoted.
     */
```

This is a little embarrassing here since the PromotionCondition does not have a dedicated context for exposing errors to the end user (beside logs).